File: ios/Sources/InAppBrowserPlugin/OSInAppBrowserError.swift
enum OSInAppBrowserTarget {
    case externalBrowser
    case systemBrowser
    case webView
}

enum OSInAppBrowserError: Error {
    case bridgeNotInitialised
    case invalidURLScheme
    case inputArgumentsIssue(target: OSInAppBrowserTarget)
    case failedToOpen(url: String, onTarget: OSInAppBrowserTarget)
    case noBrowserToClose

    var description: String {
        let result: String

        switch self {
        case .bridgeNotInitialised:
            result = "Capacitor bridge isn't initialized."
        case .invalidURLScheme:
            result = "The URL provided must begin with either http:// or https://."
        case .inputArgumentsIssue(let target):
            let targetString = switch target {
            case .externalBrowser: "openInExternalBrowser"
            case .systemBrowser: "openInSystemBrowser"
            case .webView: "openInWebView"
            }

            result = "The '\(targetString)' input parameters aren't valid."
        case .failedToOpen(url: let url, onTarget: let target):
            let targetString = switch target {
            case .externalBrowser: "External browser"
            case .systemBrowser: "SafariViewController"
            case .webView: "The WebView"
            }

            result = "\(targetString) couldn't open the following URL: '\(url)'"
        case .noBrowserToClose:
            result = "There's no browser view to close."
        }

        return result
    }

    var code: String {
        let baseCode: Int
        switch self {
        case .invalidURLScheme:
            baseCode = 3
        case .inputArgumentsIssue(let target):
            baseCode = switch target {
            case .externalBrowser: 5
            case .systemBrowser: 6
            case .webView: 7
            }
        case .failedToOpen(_, let target):
            baseCode = switch target {
            case .externalBrowser: 8
            case .systemBrowser: 9
            case .webView: 11
            }
        case .noBrowserToClose:
            baseCode = 12
        case .bridgeNotInitialised:
            baseCode = 13
        }
        return "OS-PLUG-IABP-\(String(format: "%04d", baseCode))"
    }

}

File: android/src/main/java/com/capacitorjs/osinappbrowser/OSInAppBrowserError.kt
package com.capacitorjs.osinappbrowser

sealed class OSInAppBrowserError(val code: String, val message: String) {
    data class InputArgumentsIssue(val target: OSInAppBrowserTarget) : OSInAppBrowserError(
        code = target.inputIssueCode.formatErrorCode(),
        message = "The '${target.inputIssueText}' input parameters aren't valid."
    )

    data class OpenFailed(val url: String, val target: OSInAppBrowserTarget) : OSInAppBrowserError(
        code = target.openFailedCode.formatErrorCode(),
        message = "${target.openFailedText} couldn't open the following URL: $url"
    )

    data object CloseFailed : OSInAppBrowserError(
        code = 12.formatErrorCode(),
        message = "There's no browser view to close."
    )

    data object InvalidURL : OSInAppBrowserError(
        code = 3.formatErrorCode(),
        message = "The URL provided must begin with either http:// or https://."
    )

    data class GeneralError(val url: String, val target: OSInAppBrowserTarget, val exception: Exception) : OSInAppBrowserError(
        code = target.generalErrorCode.formatErrorCode(),
        message = "An error occurred while trying to open the '${target.openFailedText}': ${exception.message}"
    )
}

enum class OSInAppBrowserTarget(
    val inputIssueCode: Int,
    val inputIssueText: String,
    val openFailedCode: Int,
    val openFailedText: String,
    val generalErrorCode: Int
) {
    EXTERNAL_BROWSER(5, "openInExternalBrowser", 8, "External browser", 13),
    SYSTEM_BROWSER(6, "openInSystemBrowser", 10, "Custom Tabs", 14),
    WEB_VIEW(7, "openInWebView", 11, "The WebView", 15)
}

private fun Int.formatErrorCode(): String {
    return "OS-PLUG-IABP-" + this.toString().padStart(4, '0')
}